src/App.tsx

```diff
@@ -1,24 +1,276 @@
+
+      return;
+    }
+
+    const normalizedTitle = title.trim();
+
+    const newTodo = {
+      title: normalizedTitle,
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setTempTodo({
+      ...newTodo,
+      id,
+    });
+
+    try {
+      const addingTodo = await addTodo(newTodo);
+
+      setTodos(prevTodos => ([
+        ...prevTodos,
+        addingTodo,
+      ]));
+    } catch {
+      handleError(ErrorsType.ADD);
+    } finally {
+      setTempTodo(null);
+    }
+  }, []);
+
+  const handlerSortType = useCallback((type: FilterType) => {
+    setSortType(type);
+  }, []);
+
+  const handleCloseError = () => {
+    setIsHiddenError(true);
+  };
+
+  const handleDeleteTodo = useCallback(async (todoId: number) => {
+    try {
+      setLoadingTodosIds(prevLoaderIds => ([
+        ...prevLoaderIds,
+        todoId,
+      ]));
+
+      await removeTodo(todoId);
+
+      setTodos(prevTodos => prevTodos.filter(todo => todoId !== todo.id));
+    } catch {
+      handleError(ErrorsType.DELETE);
+    } finally {
+      setLoadingTodosIds(prevLoaderIds => (
+        prevLoaderIds.filter(id => id !== todoId)
+      ));
+    }
+  }, []);
+
+  const handleRemoveAllCompletedTodo = useCallback(async () => {
+    const completedTodosIds = todos
+      .filter(todo => todo.completed)
+      .map(todo => todo.id);
+
+    await Promise.all(completedTodosIds.map(handleDeleteTodo));
+  }, [todos]);
+
+  const handlerTodoToggle = useCallback(async (
+    id: number,
+    completed: boolean,
+  ) => {
+    try {
+      setLoadingTodosIds(prevLoaderIds => ([
+        ...prevLoaderIds,
+        id,
+      ]));
+
+      await changeTodo(id, { completed: !completed });
+
+      setTodos(prevTodos => prevTodos.map(todo => {
+        if (todo.id === id) {
+          return { ...todo, completed: !completed };
+        }
+
+        return todo;
+      }));
+    } catch {
+      handleError(ErrorsType.UPDATE);
+    } finally {
+      setLoadingTodosIds(prevLoaderIds => (
+        prevLoaderIds.filter(prevLoaderId => id !== prevLoaderId)
+      ));
+    }
+  }, []);
+
+  const handlerToggleAll = useCallback(async () => {
+    const areAllCompleted = todos.every(todo => todo.completed);
+    const todosIds = areAllCompleted
+      ? todos.map(todo => todo.id)
+      : todos.filter(todo => !todo.completed).map(todo => todo.id);
+
+    await Promise.all(
+      todosIds.map(id => handlerTodoToggle(id, areAllCompleted)),
+    );
+  }, [todos]);
+
+  const handlerChangeTitle = useCallback(async (
+    newTitle: string,
+    todoId: number,
+  ) => {
+    const title = newTitle.trim();
+
+    if (!title) {
+      await handleDeleteTodo(todoId);
+
+      return;
+    }
+
+    try {
+      setLoadingTodosIds(prevLoadIds => ([
+        ...prevLoadIds,
+        todoId,
+      ]));
+
+      await changeTodo(todoId, { title });
+
+      setTodos(prevTodos => prevTodos.map(todo => {
+        if (todo.id === todoId) {
+          return { ...todo, title: newTitle };
```

Comment:
1 - you sent the trimmed title to API in `changeTodo`, but you set not trimmed `newTitle` fo state as current value
2 - use values that are returned from API, it can easily be changed on API, so better to use it as a source of truth